File: example/src/main/java/com/vimeo/example/vimeonetworking/AndroidGsonDeserializer.java
package com.vimeo.example.vimeonetworking;

import android.os.AsyncTask;

import com.google.gson.Gson;
import com.vimeo.networking.GsonDeserializer;
import com.vimeo.networking.callbacks.ModelCallback;

/**
 * Created by zetterstromk on 7/24/15.
 */
public class AndroidGsonDeserializer extends GsonDeserializer {

    @Override
    public void deserialize(Gson gson, Object object, ModelCallback callback) {
        new DeserializeTask(gson, object, callback).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    private class DeserializeTask extends AsyncTask<Void, Void, Object> {

        private Gson gson;
        private Object object;
        private ModelCallback callback;

        public DeserializeTask(Gson gson, Object object, ModelCallback callback) {
            this.gson = gson;
            this.object = object;
            this.callback = callback;
        }

        @Override
        protected Object doInBackground(Void... params) {
            return deserializeObject(gson, object, callback);
        }

        @Override
        protected void onPostExecute(Object result) {
            callback.success(result);
        }
    }
}

File: example-kotlin/src/main/kotlin/com/vimeo/android/networking/example/kotlin/vimeonetworking/AndroidGsonDeserializer.kt
package com.vimeo.android.networking.example.kotlin.vimeonetworking

import android.os.AsyncTask
import com.google.gson.Gson
import com.vimeo.networking.GsonDeserializer
import com.vimeo.networking.callbacks.ModelCallback

/**
 * Created by zetterstromk on 7/24/15.
 */
class AndroidGsonDeserializer : GsonDeserializer() {


    override fun deserialize(gson: Gson, any: Any, callback: ModelCallback<Any>) {
        DeserializeTask(gson, any, callback).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
    }

    private inner class DeserializeTask(private val gson: Gson,
                                        private val any: Any,
                                        private val callback: ModelCallback<Any>) : AsyncTask<Void, Void, Any>() {

        override fun doInBackground(vararg params: Void): Any? {
            return deserializeObject(gson, any, callback)
        }

        override fun onPostExecute(result: Any) {
            callback.success(result)
        }
    }
}
